### ISSUE TITLE
GRU/LSTM stateful implementation or workaround guidance

### ISSUE THREAD

**Issue by aawce** [microsoft/onnxruntime#6936]
Hi,

We are doing real-time classification and very much like onnxruntime - been trying it for some time now (v1.7.1)

However, we need to implement stateful=true in our GRU and LSTMs. We understand that onnx doesn't support stateful, and we have a way of specifying stateful via metadata.

We have reviewed other threads suggesting to read the state after a session.run, and pass it back in to the next session.run. This might work for some cases, but it's not clear if it works for some of our use cases where we may want to run batches of sequences/timesteps.

However a thought (option2) is to copy the cpu-provider GRU into our own EP and modify it to keep state. This would be our preferred way to do it BTW for several other reasons, but we'd need help and guidance to do this.

So Q to ort team - is option2 a recommended approach for early development of this feature, and if so is an expert available to spend a little time guiding us on how to do it? We are willing to roll up our sleeves to do at much as possible, assuming we haven't oversimplified the effort required.

Thks,
Chris

**Comment by skottmckay** [comment]
Why wouldn't it work if there were batches? GRU and LSTM supports batches. The GRU Y_h output and initial_h inputs have a batch size dimension, as does the LSTM Y_h/Y_c outputs and initial_h/initial_c inputs. I'd strongly suggest establishing whether this can satisfy your requirements as a first step.

It may be simpler to create a custom operator instead of a complete EP. There's documentation on doing so here: https://www.onnxruntime.ai/docs/how-to/add-custom-op.html. 
You'd have to edit the model to change the operator name of the GRU node to your custom operator name, but that should be fairly trivial to do via a simple python script. 

If you decide to do a custom op or EP, bear in mind that the InferenceSession's Run method expects execution to be stateless, so you may need to protect against concurrent usage within your op implementation if that is a potential issue.

**Comment by aawce** [comment]
Thks - our tests show that for a batch of sequences/timesteps ort will reset state between tsteps, which make sense since that's our understanding of non-stateful behavior in Keras, etc.  So when we run a batch, we want the init state of the previous tstep to be used in the next tstep within that batch.  That doesn't seem possible even with GRU_Y_h output has batch_size dimension.

So it looks fairly simple to create a blank EP,  but its unclear to me how to add an operator as the cpu-provider examples seems very complicated.  What do you recommend as the easiest operator to add to an EP?  Is there a very simple example to reference?  I've looked through cpu-provider and armnn provider and those have lots of dependencies and code to sift through.  Is there something simpler?

**Comment by skottmckay** [comment]
Can you clarify 'the init state of the previous tstep to be used in the next'? The output state of the previous step is used in the next (you can get the output state from every step via the Y output) - the Y_h output is just the state at end . Wouldn't using the init state mean using the same value for all steps?

**Comment by aawce** [comment]
So for example, with batch size = 8,  tstep = 16,  we open a session, and run a data set, internally ort will run the first 16 tsteps, then reset state and the 2nd batch of 16 tsteps, etc.  We have no way of intercepting between the batches to save and restore state.  When running this for real time detection we'd want to maintain state (recursive system can have a longer impulse response).
Anyway, it's more important we get our own EP with custom GRU that can maintain state.  Would you be able to help us with that?

**Comment by skottmckay** [comment]
If you still wish to add a custom EP, the documentation is at: https://www.onnxruntime.ai/docs/how-to/add-execution-provider.html

> We have no way of intercepting between the batches to save and restore state.

If you make the Y_h output from the GRU node a model output, and you provide the initial_h input of the GRU node as model input, you essentially intercept the state and can pass it into the next run. 

Let's say you had a really simple model. 
  - Model contains a GRU node followed by a Softmax.
  - Let's call the model input 'InputA', and it has shape {seq_length, batch, input_size}. We use this as the X input to the GRU node.
  - Let's say the GRU node Y_h output is called 'GRU_state', and that's the input to the Softmax node.
  - Let's say the model output, which comes from the Softmax node, is called 'OutputA'.

With this setup, the GRU uses the default value of initial_h which is all zeros (i.e. the state is 'reset'), and 'GRU_state' only goes into Softmax so is not accessible.

We first want to capture the state, so we add the 'GRU_state' output from the GRU node to the list of model outputs (the name just needs to be added to the output list). Now 'GRU_state' goes to the Softmax node, and is also available as a model output.

To use that state in the next call, we add a model input for this value (let's call it 'InputState'), and connect it to the initial_h input of the GRU node. 

Now that we're not using the default value of initial_h in the GRU node, for the first call we need to provide an input filled with zeros as 'InputState' with shape {num_directions, batch_size, hidden_size}.
At the end of the first call we read 'OutputA' and 'GRU_state' from the model outputs, and pass 'GRU_state' in as 'InputState' when running the next batch..

**Comment by aawce** [comment]
Hi Scott,

Thks - we have our own EP started, but just a shell at this point.  Do you know anyone who could help us copy the GRU into our EP and make these modifications?

Thks,
Chris

**Comment by skottmckay** [comment]
> Thks - we have our own EP started, but just a shell at this point. Do you know anyone who could help us copy the GRU into our EP and make these modifications?

You're probably on your own for that given your desired approach doesn't really fit with ONNX Runtime's 'run' being stateless.

As ONNX Runtime is open source you're free to copy the GRU code and modify it as needed (assuming whatever necessary license info is copied across to make the lawyers happy).

> This might work for some cases, but it's not clear if it works for some of our use cases where we may want to run batches of sequences/timesteps.

Out of interest, what is the reason why passing the state between calls to Run doesn't work for you? I'd just like to understand why the overhead of creating a new EP (or custom op) is required.
